[Woo POS][Product Search] Send Name and SKU search parameter

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOMOB-260
Merge after: #15524 
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR sends the search term in the `search_name_or_sku` field as well as the `search` field.

At the moment, the API doesn't implement this new field – in future, it'll be used in preference to the `search` parameter. When users upgrade to the future version of WooCommerce that implements it, they'll be able to get results for matches of their search term in the name or SKU fields.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Enable the `searchProductsInPOS` feature flag and set up a network proxy tool (or use Wormholy.)

1. Launch the app and navigate to POS
2. Tap the search box and type a search term
3. Inspect the network request made, observe that they send the search term in both `search` and `search_name_and_sku` fields.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I've tested this on an iPad Air running iOS 17.7.3

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
![Request log from Proxyman showing the search and search_name_or_sku parameters highlighted](https://github.com/user-attachments/assets/4864af5c-0ce4-4d13-aac7-66a434ba0532)

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
Send name or SKU search parameter to API

iamgabrielma reviewed (approved):
Works well! :shipit: 

Not sure if you've already implemented or thought about debouncing at this point, but at some points seems to make the requests too close to eachother:
``` 
09:50:48.043 &search_name_or_sku=&status=publish&type=simple
09:50:50.744 &search=Sh&search_name_or_sku=Sh&status=publish&type=simple
09:50:50.768 &search=Sushi&search_name_or_sku=Sushi&status=publish&type=simple
```

joshheald commented:
> Works well! :shipit:
> 
> Not sure if you've already implemented or thought about debouncing at this point, but at some points seems to make the requests too close to eachother:
> 
> ```
> 09:50:48.043 &search_name_or_sku=&status=publish&type=simple
> 09:50:50.744 &search=Sh&search_name_or_sku=Sh&status=publish&type=simple
> 09:50:50.768 &search=Sushi&search_name_or_sku=Sushi&status=publish&type=simple
> ```

Thanks 😊 

There is debouncing, to 300ms. The timings in your logs don't look quite what we're looking for, but the first search is deliberately not debounced. Also, if it returns before the debounce time elapses, we send the next keystroke immediately anyway. This is all to make sure it still feels responsive, but we can tweak based on feedback.

Thank you for including the logs; I've just noticed that the `search_name_or_sku` is sent empty in the first request and it shouldn't be sent at all. I'll fix that before merging.

joshheald committed:
Don’t search when the term has been deleted

This view shows the pre-search UI if there’s no search term, so we don’t need to send this network request, we can just act as if it was completed.

joshheald commented:
> Thank you for including the logs; I've just noticed that the `search_name_or_sku` is sent empty in the first request and it shouldn't be sent at all. I'll fix that before merging.

It looks like `""` is only sent as a search term when you delete all the text from the field. I've prevented that one in the view layer in 60d00f4.

joshheald committed:
Merge branch 'woomob-315-pre-search-load-popular-products' into woomob-360-send-sku-search-parameter
